rust/module/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+///     description: b"My very own kernel module!",
+///     license: b"GPL v2",
+///     params: {},
+/// }
+///
+/// struct MyKernelModule;
+///
+/// impl KernelModule for MyKernelModule {
+///     fn init() -> KernelResult<Self> {
+///         Ok(MyKernelModule)
+///     }
+/// }
+/// ```
+#[proc_macro]
+pub fn module(ts: TokenStream) -> TokenStream {
+    let mut it = ts.into_iter();
+
+    let type_ = get_ident(&mut it, "type");
+    let name = get_byte_string(&mut it, "name");
+    let author = get_byte_string(&mut it, "author");
+    let description = get_byte_string(&mut it, "description");
+    let license = get_byte_string(&mut it, "license");
+    let params = get_group(&mut it, "params");
+
+    expect_end(&mut it);
+
+    assert_eq!(params.delimiter(), Delimiter::Brace);
+
+    let mut it = params.stream().into_iter();
+
+    let mut params_modinfo = String::new();
+
+    loop {
+        let param_name = match it.next() {
+            Some(TokenTree::Ident(ident)) => ident.to_string(),
+            Some(_) => panic!("Expected Ident or end"),
+            None => break,
+        };
+
+        assert_eq!(expect_punct(&mut it), ':');
+        let param_type = expect_ident(&mut it);
+        let group = expect_group(&mut it);
+        assert_eq!(expect_punct(&mut it), ',');
+
+        assert_eq!(group.delimiter(), Delimiter::Brace);
+
+        let mut param_it = group.stream().into_iter();
+        let param_default = if param_type == "bool" {
+            get_ident(&mut param_it, "default")
+        } else {
+            get_literal(&mut param_it, "default")
+        };
+        let param_permissions = get_literal(&mut param_it, "permissions");
+        let param_description = get_byte_string(&mut param_it, "description");
+        expect_end(&mut param_it);
+
+        // TODO: more primitive types
+        // TODO: other kinds: arrays, unsafes, etc.
+        let param_kernel_type = match param_type.as_ref() {
+            "bool" => "bool",
+            "i32" => "int",
+            t => panic!("Unrecognized type {}", t),
+        };
+
+        params_modinfo.push_str(&build_modinfo_string_param(&name, "parmtype", &param_name, &param_kernel_type));
+        params_modinfo.push_str(&build_modinfo_string_param(&name, "parm", &param_name, &param_description));
+        params_modinfo.push_str(
+            &format!(
+                "
+                static mut __{name}_{param_name}_value: {param_type} = {param_default};
+
+                struct __{name}_{param_name};
+
+                impl __{name}_{param_name} {{
+                    fn read(&self) -> {param_type} {{
+                        unsafe {{ __{name}_{param_name}_value }}
+                    }}
+                }}
+
+                const {param_name}: __{name}_{param_name} = __{name}_{param_name};
+
+                // FIXME: does the `align` do the right thing here?
+                // `core::mem::size_of(usize)`
+                #[repr(C,align(8))]
```

Comment:
Should this just be `#[repr(transparent)]` so it has the same alignment as the `kernel_param`?

Reply:
This question took me down a rabbit hole... :-)

I was following the C macro (`__module_param_call`) which declares the static variable in the `__param` section as `aligned(sizeof(void *))`. Digging a bit why in the historical repos, this goes back to [a 2003 fix from Andi Kleen](https://git.kernel.org/pub/scm/linux/kernel/git/tglx/history.git/commit/include/linux/moduleparam.h?id=38d5b085d2a0d36ece55bcede84be9ae88cd4481):

> The problem seems to be that gcc generates a 32byte alignment for static
objects > 32bytes.  This causes gas to set a high alignment on the
section, which causes the uneven (not multiple of sizeof(struct
kernel_param)) section size.  The pointer division with a base not being
a multiple of sizeof(*ptr) then causes the invalid result.
>
>This just forces a small alignment, which makes the section end come out
with the correct alignment.
>
> The only mystery left is why ld chose a 16 byte padding instead of
32byte.

Given this came up during the GCC 3.x era, it is likely it is fixed (I can't reproduce it either), so it is likely we can remove it from the kernel itself (unless somehow some architecture depends on that nowadays). I will ask around. For the moment both ways work -- I will simplify with `transparent` but put some info in the comment to keep it in mind.